ttg/ttg/make_tt.h

```diff
@@ -329,24 +357,46 @@ auto make_tt_tpl(funcT &&func, const std::tuple<ttg::Edge<keyT, input_edge_value
-/// @param[in] func a callable object; if `ttg::meta::is_void_v<keyT>==true`, the signature
-///         must be `void(input_valuesT&&..., std::tuple<output_terminalsT...>&)`,
-///         else `void(const keyT&, input_valuesT&&..., std::tuple<output_terminalsT...>&)`
+/// @param[in] func a callable object; it can be _generic_ (e.g., a template function, a generic lambda, etc.; see
+///            below) or _nongeneric_ (with concrete types for its arguments). In either case its signature
+///            must match the following:
+///         - if `ttg::meta::is_void_v<keyT>==true`:
+///           - `void(input_valuesT&&..., std::tuple<output_terminalsT...>&)`: full form, with the explicitly-passed
+///             output terminals ensuring compile-time type-checking of the dataflow into the output terminals (see ttg::send);
+///           - `void(input_valuesT&&...)`: simplified form, with no type-checking of the dataflow into the output terminals;
+///         - if `ttg::meta::is_void_v<keyT>==false`:
+///           - `void(const keyT&, input_valuesT&&..., std::tuple<output_terminalsT...>&)`: full form, with the explicitly-passed
+/////             output terminals ensuring compile-time type-checking of the dataflow into the output terminals (see ttg::send);
+///           - `void(const keyT&, input_valuesT&&...)`: simplified form, with no type-checking of the dataflow into the output terminals.
 /// @param[in] inedges a tuple of input edges
 /// @param[in] outedges a tuple of output edges
 /// @param[in] name a string label for the resulting TT
 /// @param[in] name a string label for the resulting TT
 /// @param[in] innames string labels for the respective input terminals of the resulting TT
 /// @param[in] outnames string labels for the respective output terminals of the resulting TT
 ///
-/// @internal To be able to handle generic callables the input edges are used to determine the trial set of
-/// argument types.
+/// @note For generic callables the arguments that are used read-only should be declared as `U&` (where `U` is the corresponding template parameter)
+///       or `auto&` (in contexts such as generic lambdas where template arguments are implicit). The arguments that are
+///       to be consumed (e.g. mutated, moved, etc.) should be declared as `U&&` or `auto&&` (i.e., as universal references).
+///       For example, in
+///       @code
+///          make_tt([](auto& key, auto& datum1, auto&& datum2) { ... }, ...);
+///       @endcode
+///       the task id (`key`) and the first datum will be passed by const lvalue reference (i.e. no copy will be created by the runtime),
+///       whereas the second datum will be passed by an rvalue reference, which may cause copying.
+///       The corresponding free function analog of the above lambda is:
+///       @code
+///          template <typename K, typename D1, typename D2>
+///          void func (K& key, D1& datum1, D2&& datum2) { ... }
```

Comment:
Shouldn't that be `void func (K& key, const D1& datum1, D2&& datum2) { ... }` because `datum1` is `auto&`?

Reply:
no. operationally just think of `f(auto&)` as `template <typename T> f(T&)` ... so `auto&` becomes `D1&`, not `const D1&`, e.g.

```
  template <typename K, typename D1, typename D2>
  void func0(K& key, D1& datum1, D2&& datum2) {
    abort();
  }

// OK
static_assert(std::is_void_v<decltype(func0(std::declval<const int&>(), std::declval<const float &>(), std::declval<const float &>()))>);
// error: no matching function for call to 'func0', candidate function [with K = const int, D1 = float, D2 = float] not viable: expects an lvalue for 2nd argument
static_assert(std::is_void_v<decltype(tt_i_iv::func0(std::declval<const int&>(), std::declval<float &&>(), std::declval<float &&>()))>);
```